microprofile/fault-tolerance-smallrye/extension/src/main/java/org/wildfly/extension/microprofile/faulttolerance/deployment/MicroProfileFaultToleranceDeploymentProcessor.java

```diff
@@ -0,0 +1,104 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.wildfly.extension.microprofile.faulttolerance.deployment;
+
+import static org.jboss.as.weld.Capabilities.WELD_CAPABILITY_NAME;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.netflix.config.ConfigurationManager;
+import org.apache.commons.configuration.AbstractConfiguration;
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.jboss.as.controller.capability.CapabilityServiceSupport;
+import org.jboss.as.server.deployment.AttachmentKey;
+import org.jboss.as.server.deployment.Attachments;
+import org.jboss.as.server.deployment.DeploymentPhaseContext;
+import org.jboss.as.server.deployment.DeploymentUnit;
+import org.jboss.as.server.deployment.DeploymentUnitProcessingException;
+import org.jboss.as.server.deployment.DeploymentUnitProcessor;
+import org.jboss.as.weld.Capabilities;
+import org.jboss.as.weld.WeldCapability;
+import org.jboss.modules.Module;
+import org.wildfly.extension.microprofile.faulttolerance.MicroProfileFaultToleranceLogger;
+
+/**
+ * @author Radoslav Husar
+ */
+public class MicroProfileFaultToleranceDeploymentProcessor implements DeploymentUnitProcessor {
+
+    private static final AttachmentKey<Boolean> ATTACHMENT_KEY = AttachmentKey.create(Boolean.class);
+    private static final AtomicInteger remainingHystrixConfiguringDeployments = new AtomicInteger();
+
+    @Override
+    public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitProcessingException {
+        DeploymentUnit deploymentUnit = phaseContext.getDeploymentUnit();
+
+        // Weld Extension
+        CapabilityServiceSupport support = deploymentUnit.getAttachment(Attachments.CAPABILITY_SERVICE_SUPPORT);
+
+        WeldCapability weldCapability;
+        try {
+            weldCapability = support.getCapabilityRuntimeAPI(Capabilities.WELD_CAPABILITY_NAME, WeldCapability.class);
+        } catch (CapabilityServiceSupport.NoSuchCapabilityException e) {
+            throw MicroProfileFaultToleranceLogger.ROOT_LOGGER.deploymentRequiresCapability(deploymentUnit.getName(), WELD_CAPABILITY_NAME);
+        }
+        if (weldCapability.isPartOfWeldDeployment(deploymentUnit)) {
+            weldCapability.registerExtensionInstance(new MicroProfileFaultToleranceCDIExtension(), deploymentUnit);
+        }
+
+        synchronized (remainingHystrixConfiguringDeployments) {
+            if (remainingHystrixConfiguringDeployments.get() == 0) {
```

Comment:
This is racy.  Any thread reaching this block will compete with a undeploy thread that just set the remainingHystrixConfiguringDeployments counter to 0.  This can result an undeploy thread clearing the configuration that a deploy thread just created.

Reply:
Good point, org.wildfly.extension.microprofile.faulttolerance.deployment.MicroProfileFaultToleranceDeploymentProcessor#undeploy needs to be in the synchronized block as well then.

Reply:
You could use: 
```java
if (remainingHystrixConfiguringDeployments.compareAndSet(0, 1)) {
   // create config
} else {
   remainingHystrixConfiguringDeployments.incrementAndGet();
}
```

In which case, you don't need to synchronize anything - however, this might allow deployments to attempt to read the config before it is finished being created.  You could then solve that problem with a latch or barrier.  However, it might just be easier to use synchronized blocks in deploy vs undeploy, in which case you can replace the AtomicInteger with a volatile int.

Reply:
Using synchronized block now.